## Practice Exercise 2

In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to import the movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [ ]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

In [ ]:
import os
current_directory = os.getcwd()
print(f"Current working directory: {current_directory}")

Current working directory: /content


In [ ]:
os.chdir('/content/drive/MyDrive/Colab Notebooks/AICode/AICode586/Pandas_Numpy')
current_directory = os.getcwd()
print(f"Current working directory: {current_directory}")

Current working directory: /content/drive/MyDrive/Colab Notebooks/AICode/AICode586/Pandas_Numpy


### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [ ]:
# Write your code for importing the csv file here
movies = pd.read_csv('Movies.csv')
movies

,color,director_name,num_critic_for_reviews,duration,director_facebook_likes,actor_3_facebook_likes,actor_2_name,actor_1_facebook_likes,gross,genres,...,num_user_for_reviews,language,country,content_rating,budget,title_year,actor_2_facebook_likes,imdb_score,aspect_ratio,movie_facebook_likes
0,Color,James Cameron,723.0,178.0,0.0,855.0,Joel David Moore,1000.0,760505847.0,Action|Adventure|Fantasy|Sci-Fi,...,3054.0,English,USA,PG-13,237000000.0,2009.0,936.0,7.9,1.78,33000
1,Color,Gore Verbinski,302.0,169.0,563.0,1000.0,Orlando Bloom,40000.0,309404152.0,Action|Adventure|Fantasy,...,1238.0,English,USA,PG-13,300000000.0,2007.0,5000.0,7.1,2.35,0
2,Color,Sam Mendes,602.0,148.0,0.0,161.0,Rory Kinnear,11000.0,200074175.0,Action|Adventure|Thriller,...,994.0,English,UK,PG-13,245000000.0,2015.0,393.0,6.8,2.35,85000
3,Color,Christopher Nolan,813.0,164.0,22000.0,23000.0,Christian Bale,27000.0,448130642.0,Action|Thriller,...,2701.0,English,USA,PG-13,250000000.0,2012.0,23000.0,8.5,2.35,164000
4,Color,Andrew Stanton,462.0,132.0,475.0,530.0,Samantha Morton,640.0,73058679.0,Action|Adventure|Sci-Fi,...,738.0,English,USA,PG-13,263700000.0,2012.0,632.0,6.6,2.35,24000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3848,Color,Shane Carruth,143.0,77.0,291.0,8.0,David Sullivan,291.0,424760.0,Drama|Sci-Fi|Thriller,...,371.0,English,USA,PG-13,7000.0,2004.0,45.0,7.0,1.85,19000
3849,Color,Neill Dela Llana,35.0,80.0,0.0,0.0,Edgar Tancangco,0.0,70071.0,Thriller,...,35.0,English,Philippines,Not Rated,7000.0,2005.0,0.0,6.3,NaN,74
3850,Color,Robert Rodriguez,56.0,81.0,0.0,6.0,Peter Marquardt,121.0,2040920.0,Action|Crime|Drama|Romance|Thriller,...,130.0,Spanish,USA,R,7000.0,1992.0,20.0,6.9,1.37,0
3851,Color,Edward Burns,14.0,95.0,0.0,133.0,Caitlin FitzGerald,296.0,4584.0,Comedy|Drama,...,14.0,English,USA,Not Rated,9000.0,2011.0,205.0,6.4,NaN,413


**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [ ]:
# Write your code for inspection here
movies.head()

,color,director_name,num_critic_for_reviews,duration,director_facebook_likes,actor_3_facebook_likes,actor_2_name,actor_1_facebook_likes,gross,genres,...,num_user_for_reviews,language,country,content_rating,budget,title_year,actor_2_facebook_likes,imdb_score,aspect_ratio,movie_facebook_likes
0,Color,James Cameron,723.0,178.0,0.0,855.0,Joel David Moore,1000.0,760505847.0,Action|Adventure|Fantasy|Sci-Fi,...,3054.0,English,USA,PG-13,237000000.0,2009.0,936.0,7.9,1.78,33000
1,Color,Gore Verbinski,302.0,169.0,563.0,1000.0,Orlando Bloom,40000.0,309404152.0,Action|Adventure|Fantasy,...,1238.0,English,USA,PG-13,300000000.0,2007.0,5000.0,7.1,2.35,0
2,Color,Sam Mendes,602.0,148.0,0.0,161.0,Rory Kinnear,11000.0,200074175.0,Action|Adventure|Thriller,...,994.0,English,UK,PG-13,245000000.0,2015.0,393.0,6.8,2.35,85000
3,Color,Christopher Nolan,813.0,164.0,22000.0,23000.0,Christian Bale,27000.0,448130642.0,Action|Thriller,...,2701.0,English,USA,PG-13,250000000.0,2012.0,23000.0,8.5,2.35,164000
4,Color,Andrew Stanton,462.0,132.0,475.0,530.0,Samantha Morton,640.0,73058679.0,Action|Adventure|Sci-Fi,...,738.0,English,USA,PG-13,263700000.0,2012.0,632.0,6.6,2.35,24000


In [ ]:
movies.describe()

,num_critic_for_reviews,duration,director_facebook_likes,actor_3_facebook_likes,actor_1_facebook_likes,gross,num_voted_users,cast_total_facebook_likes,facenumber_in_poster,num_user_for_reviews,budget,title_year,actor_2_facebook_likes,imdb_score,aspect_ratio,movie_facebook_likes
count,3852.000000,3852.000000,3853.000000,3847.000000,3853.000000,3.853000e+03,3.853000e+03,3853.000000,3847.000000,3853.000000,3.853000e+03,3853.000000,3852.000000,3853.000000,3781.000000,3853.000000
mean,163.046989,109.921080,784.507397,747.290096,7575.558526,5.096547e+07,1.024240e+05,11239.480145,1.378217,326.720478,4.524817e+07,2003.061511,1959.018432,6.462886,2.109569,9089.546068
std,123.927258,22.741728,3027.395958,1841.623945,15405.383114,6.932033e+07,1.502942e+05,18922.948757,2.056071,407.920904,2.234208e+08,10.007168,4472.171290,1.053843,0.353236,21277.375223
min,1.000000,34.000000,0.000000,0.000000,0.000000,1.620000e+02,5.000000e+00,0.000000,0.000000,1.000000,2.180000e+02,1920.000000,0.000000,1.600000,1.180000,0.000000
25%,72.000000,95.000000,10.000000,183.000000,721.000000,6.830957e+06,1.730900e+04,1817.000000,0.000000,102.000000,1.000000e+07,1999.000000,362.000000,5.900000,1.850000,0.000000
50%,134.000000,106.000000,58.000000,427.000000,1000.000000,2.790000e+07,5.052300e+04,3876.000000,1.000000,203.000000,2.400000e+07,2005.000000,664.000000,6.600000,2.350000,206.000000
75%,221.000000,120.000000,222.000000,685.000000,12000.000000,6.550000e+07,1.241850e+05,15972.000000,2.000000,391.000000,5.000000e+07,2010.000000,971.000000,7.200000,2.350000,11000.000000
max,813.000000,330.000000,23000.000000,23000.000000,640000.000000,7.605058e+08,1.689764e+06,656730.000000,43.000000,5060.000000,1.221550e+10,2016.000000,137000.000000,9.300000,16.000000,349000.000000


In [ ]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3853 entries, 0 to 3852
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      3851 non-null   object 
 1   director_name              3853 non-null   object 
 2   num_critic_for_reviews     3852 non-null   float64
 3   duration                   3852 non-null   float64
 4   director_facebook_likes    3853 non-null   float64
 5   actor_3_facebook_likes     3847 non-null   float64
 6   actor_2_name               3852 non-null   object 
 7   actor_1_facebook_likes     3853 non-null   float64
 8   gross                      3853 non-null   float64
 9   genres                     3853 non-null   object 
 10  actor_1_name               3853 non-null   object 
 11  movie_title                3853 non-null   object 
 12  num_voted_users            3853 non-null   int64  
 13  cast_total_facebook_likes  3853 non-null   int64

#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  <font color='red'>(3821, 26)</font>
-  <font color='red'>(3879, 28)</font>
-  <font color='red'>(3853, 28)</font>
-  <font color='red'>(3866, 26)</font>

In [ ]:
# Write you code here

movies.shape

(3853, 28)

#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font>

In [ ]:
# Write you code here

null_counts = movies.isnull().sum()
num_columns_with_nulls = (null_counts > 0).sum()
print(f"Number of columns with null values: {num_columns_with_nulls}")

Number of columns with null values: 12


### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  `color`
-  `director_facebook_likes`
-  `actor_1_facebook_likes`
-  `actor_2_facebook_likes`
-  `actor_3_facebook_likes`
-  `actor_2_name`
-  `cast_total_facebook_likes`
-  `actor_3_name`
-  `duration`
-  `facenumber_in_poster`
-  `content_rating`
-  `country`
-  `movie_imdb_link`
-  `aspect_ratio`
-  `plot_keywords`

In [ ]:
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations

movies.drop(['color', 'director_facebook_likes', 'actor_1_facebook_likes', 'actor_2_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name', 'cast_total_facebook_likes', 'actor_3_name', 'duration', 'facenumber_in_poster', 'content_rating', 'country', 'movie_imdb_link', 'aspect_ratio', 'plot_keywords'], axis=1, inplace=True)

#### <font color='red'>Question 3: What is the count of columns in the new dataframe? </font>
-  <font color='red'>10</font>
-  <font color='red'>13</font>
-  <font color='red'>15</font>
-  <font color='red'>17</font>

In [ ]:
movies.shape

(3853, 13)

**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'.

In [ ]:
# Write you code here
null_percentages = movies.isnull().sum() * 100 / len(movies)
print(null_percentages)

director_name             0.000000
num_critic_for_reviews    0.025954
gross                     0.000000
genres                    0.000000
actor_1_name              0.000000
movie_title               0.000000
num_voted_users           0.000000
num_user_for_reviews      0.000000
language                  0.103815
budget                    0.000000
title_year                0.000000
imdb_score                0.000000
movie_facebook_likes      0.000000
dtype: float64


In [ ]:
null_percentages.index

Index(['director_name', 'num_critic_for_reviews', 'gross', 'genres',
       'actor_1_name', 'movie_title', 'num_voted_users',
       'num_user_for_reviews', 'language', 'budget', 'title_year',
       'imdb_score', 'movie_facebook_likes'],
      dtype='object')

#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='red'>language</font>
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

In [ ]:
highest_null_column = null_percentages.idxmax()
highest_null_percentage = null_percentages.max()
print(f"The column with the highest percentage of null values is '{highest_null_column}' with {highest_null_percentage:.2f}% nulls.")

The column with the highest percentage of null values is 'language' with 0.10% nulls.


**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [ ]:
# Write your code for filling the NaN values in the 'language' column here
movies['language'].fillna('English', inplace=True)

#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3674</font>
-  <font color='red'>3668</font>
-  <font color='red'>3672</font>

In [ ]:
english_movies_count = movies[movies['language'] == 'English'].shape[0]
print(f"The count of movies made in English after replacing NaN values is: {english_movies_count}")

The count of movies made in English after replacing NaN values is: 3675


### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [ ]:
# Write your code for unit conversion here
movies['budget'] = movies['budget'] / 1000000
movies['gross'] = movies['gross'] / 1000000

**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [ ]:
# Write your code for creating the profit column here
movies['profit'] = movies['gross'] - movies['budget']

In [ ]:
# Write your code for sorting the dataframe here
movies = movies.sort_values(by='profit', ascending=False)

In [ ]:
top10 = movies.head(10)

**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='red'>E.T. the Extra-Terrestrial</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

In [ ]:
top10.iloc[4]

,2748
director_name,Steven Spielberg
num_critic_for_reviews,215.0
gross,0.000435
genres,Family|Sci-Fi
actor_1_name,Henry Thomas
movie_title,E.T. the Extra-Terrestrial
num_voted_users,281842
num_user_for_reviews,515.0
language,English
budget,0.00001


In [ ]:
fifth_movie_title = top10.iloc[4]['movie_title']
print(f"The movie title of the 5th ranked movie is: {fifth_movie_title}")

The movie title of the 5th ranked movie is: E.T. the Extra-Terrestrial 


**Subtask 3.3: Find IMDb Top 250**

Create a new dataframe `IMDb_Top_250` and store the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000.

Also add a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.

In [ ]:
# Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe
# and name that dataframe as 'IMDb_Top_250'

IMDb_Top_250 = movies[movies['num_voted_users'] > 25000].sort_values(by='imdb_score', ascending=False).head(250)
IMDb_Top_250['Rank'] = range(1, 251)

In [ ]:
bins = [7.5, 8.0, 8.5, 9.0, 9.5, 10.0]
labels = ['7.5 to 8', '8 to 8.5', '8.5 to 9', '9 to 9.5', '9.5 to 10']
IMDb_Top_250['imdb_score_bucket'] = pd.cut(IMDb_Top_250['imdb_score'], bins=bins, labels=labels, right=False)
bucket_counts = IMDb_Top_250['imdb_score_bucket'].value_counts().sort_index()
print("Number of movies in each IMDb score bucket:")
print(bucket_counts)
print(f"\nThe bucket with the maximum number of movies is: {bucket_counts.idxmax()}")

Number of movies in each IMDb score bucket:
imdb_score_bucket
7.5 to 8      45
8 to 8.5     159
8.5 to 9      42
9 to 9.5       4
9.5 to 10      0
Name: count, dtype: int64

The bucket with the maximum number of movies is: 8 to 8.5


#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='red'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_250*? </font>

**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [ ]:
# Write your code for creating three new dataframes here
Meryl_Streep = movies[movies['actor_1_name'] == 'Meryl Streep']

In [ ]:
Leo_Caprio = movies[movies['actor_1_name'] == 'Leonardo DiCaprio']

In [ ]:
Brad_Pitt = movies[movies['actor_1_name'] == 'Brad Pitt']

In [ ]:
# Write your code for combining the three dataframes here
Combined = pd.concat([Meryl_Streep, Leo_Caprio, Brad_Pitt])

In [ ]:
# Write your code for grouping the combined dataframe here
actor_ratings = Combined.groupby('actor_1_name')[['num_critic_for_reviews', 'num_user_for_reviews']].mean()

In [ ]:
# Write the code for finding the mean of critic reviews and audience reviews here
print("Mean ratings for each actor:")
print(actor_ratings)

critic_favorite = actor_ratings['num_critic_for_reviews'].idxmax()
audience_favorite = actor_ratings['num_user_for_reviews'].idxmax()



Mean ratings for each actor:
                   num_critic_for_reviews  num_user_for_reviews
actor_1_name                                                   
Brad Pitt                      245.000000            742.352941
Leonardo DiCaprio              330.190476            914.476190
Meryl Streep                   181.454545            297.181818

Actor with highest mean critic reviews: Leonardo DiCaprio
Actor with highest mean user reviews: Leonardo DiCaprio


#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

In [ ]:
print(f"\nActor with highest mean critic reviews: {critic_favorite}")
print(f"Actor with highest mean user reviews: {audience_favorite}")


Actor with highest mean critic reviews: Leonardo DiCaprio
Actor with highest mean user reviews: Leonardo DiCaprio
